darshan-util/pydarshan/darshan/experimental/plots/data_access_by_filesystem.py

```diff
@@ -0,0 +1,591 @@
+    list_count_axes: list = []
+    # use log10 scale if range exceeds
+    # two orders of magnitude in a column
+    use_log = [False, False]
+    for idx, series_pair in enumerate([[bytes_rd_series, bytes_wr_series],
+                                       [file_rd_series, file_wr_series]]):
+        maxval = max(series_pair[0].max(), series_pair[1].max())
+        minval = min(series_pair[0].min(), series_pair[1].min())
+        ratio = maxval / minval
+        if ratio > 100:
+            use_log[idx] = True
+
+    for row, filesystem in enumerate(filesystem_roots):
+        ax_filesystem_bytes = fig.add_subplot(len(filesystem_roots),
+                                              2,
+                                              row * 2 + 1)
+        ax_filesystem_counts = fig.add_subplot(len(filesystem_roots),
+                                              2,
+                                              row * 2 + 2)
+        if row > 0:
+            ax_filesystem_bytes.get_shared_x_axes().join(ax_filesystem_bytes, list_byte_axes[row - 1])
+            if use_log[0]:
+                ax_filesystem_bytes.set_xscale('log')
+            ax_filesystem_counts.get_shared_x_axes().join(ax_filesystem_counts, list_count_axes[row - 1])
+            if use_log[1]:
+                ax_filesystem_counts.set_xscale('log')
+
+        list_byte_axes.append(ax_filesystem_bytes)
+        list_count_axes.append(ax_filesystem_counts)
+
+        # convert to MiB using the factor suggested
+        # by Google (approximate result only for now)
+        bytes_read = bytes_rd_series[filesystem]/1.049e+6
+        bytes_written = bytes_wr_series[filesystem]/1.049e+6
+        files_written = file_wr_series[filesystem]
+        files_read = file_rd_series[filesystem]
+
+        # scale to fit longer filesystem
+        # strings on the left side of the plots
+        # NOTE: may need more sophisticated scaling
+        # eventually
+        if len(filesystem) <= 8 and not '<STD' in filesystem:
+            fontsize = 18
+        else:
+            fontsize = 12
+
+        # anonymized STD.. streams have associated integers
+        # that are stored in the filesystem data field
+        # but that are confusing to display, so strip them
+        if filesystem.startswith('anonymized'):
+            ax_filesystem_bytes.annotate('anonymized', (-0.3, 0.5), fontsize=fontsize, xycoords='axes fraction')
+        else:
+            ax_filesystem_bytes.annotate(filesystem, (-0.3, 0.5), fontsize=fontsize, xycoords='axes fraction')
+
+        ax_filesystem_counts.barh(0, file_wr_series[filesystem], color='red', alpha=0.3)
+        ax_filesystem_counts.barh(1, file_rd_series[filesystem], color='blue', alpha=0.3)
+
+        ax_filesystem_bytes.text(0, 0.75, f' # bytes read ({bytes_read:.2E} MiB)', transform=ax_filesystem_bytes.transAxes)
+        ax_filesystem_bytes.text(0, 0.25, f' # bytes written ({bytes_written:.2E} MiB)', transform=ax_filesystem_bytes.transAxes)
```

Comment:
Generally speaking (and if it's not too much trouble), it might be nice if whatever function that generates this figure could just take a configurable number of max file systems to characterize, and it should use as much space as it needs to generate them comfortably (as opposed to them being squeezed as shown here). If there are more file systems used in the log than is allowed, just keep the top N in terms of total bytes read+written. I'm just thinking about using this sort of figure in different contexts:

1. in a web dashboard or a Jupyter notebook exploration where you want to list everything and don't care about how much space is consumed
2. in a summary report where we want to ensure that the returned figure will fit nicely in a restricted amount of canvas space, so we only want the top 5 (for instance) most important ones

Reply:
As for why there are weird //HASH file systems listed here, I had a closer look and think I understand what's going on.

As part of the log anonymization process, we anonymize the entire file path name starting at the end of the mount point into a single integer hash value (e.g., a file `/scratch/my_dir/my_file` found on mount `/scratch` might be anonymized to `/scratch/3973197598351`, throwing out the `my_dir/my_file` portion of the name). Looks like the anonymizer tool is tacking on an additional `/` resulting in things like `//2446001947` that you see above, but that shouldn't really cause any issues.

In any case, those names actually correspond to files on the `/` file system, and so they should all be collapsed into a single row. Maybe there is a slight edge case in the code for matching file names to mount points Darshan is aware of when falling through to mount `/`? The darshan-util library code should provide the array of mounts sorted in terms of decreasing path length (i.e. `/` would be the last one), so you should just need to iterate through them in order to find the first one that matches the prefix of a file you're interested in.

Reply:
For the point of the squished plots, I've added a small fix and regression test using some log repo files for that. The cramped example from `imbalanced-io.darshan` looks more reasonable for vertical scale now, and a log with less categories still seems ok (`snyder_acme.exe_id1253318_9-27-24239-1515303144625770178_2.darshan`). That said, I have not yet added an option to select the first N categories yet.

![snyder_acme exe_id1253318_9-27-24239-1515303144625770178_2 darshan png_data_access_by_category](https://user-images.githubusercontent.com/7903078/144148984-d419c317-ad46-410e-b524-4412bf45a9cd.png)
![imbalanced-io darshan png_data_access_by_category](https://user-images.githubusercontent.com/7903078/144148988-0fe51a9e-6059-42dd-9a9f-4990b84a6590.png)

Reply:
I added support for plotting the first `N` categories via a new `num_cats` argument that defaults to `None`, allowing all categories to be plotted in an expanded y dimension, or just a custom number of categories.

Cutting down `imbalanced-io.darshan` to only display a few cats (other display issues notwithstanding):

1 category:
![image](https://user-images.githubusercontent.com/7903078/144967432-ebe2c2a9-aa78-48fa-92eb-910293db2b3a.png)

3 categories:
![image](https://user-images.githubusercontent.com/7903078/144967481-506fdaba-fdbc-407d-8d58-aa0530d306be.png)

And 2 categories on the `snyder..` file:
![image](https://user-images.githubusercontent.com/7903078/144967668-43846655-fca7-4a52-a386-139176bc33de.png)

Scaling back to the full extent of `imbalanced-io.darshan` still looks "ok" I think:
![image](https://user-images.githubusercontent.com/7903078/144967753-78461441-5957-4f3a-81a8-e7e022b07a73.png)

Reply:
I still have more comments to address this in this thread... I'll leave it up to reviewers to continue going through this PR and marking as resolved those conversations that are actually resolved.

Reply:
Thanks @tylerjereddy, I think these modifications look good for the most part.

You may have been planning to get to these yet, but still think the following two things would be nice to address:

1. In the case of restricting these plots to N categories, can we pre-sort categories in decreasing order of total bytes read+written, so that we always display info on the top N most-used categories? In the `imbalanced-io.darshan` case, if I just wanted 1 category, I'd want it to be `/lus` because it's used the most by far.
2. `imbalanced-io.darshan` still displays all of the weird `//HASH` entries as mount points, but they are each individual files (you can see this in the right column, they are all just single files) that are mounted on `/` (i.e., the root file system). I can probably help figure out where the logic is being bungled here -- it could be something deeper in pydarshan than the code you've added here.